lib/src/source/hosted.dart

```diff
@@ -1100,3 +1114,95 @@ class _RefAndCache {
   @override
   bool operator ==(Object other) => other is _RefAndCache && other.ref == ref;
 }
+
+extension GCSChecksumValidation on http.StreamedResponse {
+  static const checksumsHeaderName = 'x-goog-hash';
+
+  bool get isFromGCS {
+    // TODO: Does this suffice as a heuristic?
+    return headers.containsKey(checksumsHeaderName);
+  }
+
+  /// Adds a checksum validation "tap" to the response stream and returns a
+  /// wrapped Stream object, which should be used to consume the incoming data.
+  ///
+  /// As chunks are received, a CRC32C checksum is updated.
+  /// Once the download is completed, the final checksum is compared with
+  /// the one present in the `x-goog-hash` response header from GCS.
+  ///
+  /// Throws [PackageIntegrityException] if anything is wrong with the checksum
+  /// or if there is a checksum mismatch.
+  Stream<List<int>> streamWithGCSChecksumValidationTap() {
+    if (gcsCrc32c == null) {
+      throw PackageIntegrityException(
+          'Package served from GCS is missing CRC32C checksum', request?.url);
+    }
+
+    final checksumComputer = Crc32c();
+
+    return stream.tap(checksumComputer.update, onDone: () {
+      var computedCrc32c = checksumComputer.finalize();
+
+      log.io(
+          'Computed checksum ($computedCrc32c) for package with GCS checksum of ($gcsCrc32c).');
+
+      if (gcsCrc32c != computedCrc32c) {
+        throw PackageIntegrityException(
```

Comment:
crc32c is to protect against transport issues... I think we should retry, if it fails.

Reply:
Gotcha, I'll add retries.